Source/ENS/Private/Characters/Player/EnsPlayerController.cpp

```diff
@@ -75,6 +80,85 @@ void AEnsPlayerController::Tick(const float DeltaSeconds)
 			PendingInteractObject = nullptr;
 		}
 	}
+	
+	FVector Start = ActiveCamera->GetComponentLocation();
+    FVector End = GetPawn()->GetActorLocation();
+
+    TArray<TEnumAsByte<EObjectTypeQuery>> CollisionObjectTypes;
+    CollisionObjectTypes.Add(UEngineTypes::ConvertToObjectType(ECC_WorldStatic));
+
+    TArray<AActor*> ActorsToIgnore;
+    TArray<FHitResult> OutHits;
+
+    
+
+    bool bGotHits = UKismetSystemLibrary::CapsuleTraceMultiForObjects(
+        GetWorld(), Start, End, ActiveCapsuleComponent->GetScaledCapsuleRadius(),
+        ActiveCapsuleComponent->GetScaledCapsuleHalfHeight(), CollisionObjectTypes, true,
+        ActorsToIgnore,
+        EDrawDebugTrace::None,
+        OutHits, true);
+
+	if(bGotHits)
+    {
+        TSet<UStaticMeshComponent*> MeshesJustOccluded;
+        
+        for (FHitResult Hit : OutHits)
+        {
+            const AActor* HitActor = Cast<AActor>(Hit.GetActor());
+            if (!HitActor)
+            	continue;
+
+            TArray<UStaticMeshComponent*> GroupMeshes;
+        	GetChildsOfActor(HitActor, GroupMeshes);
+            
+            for (UStaticMeshComponent* MeshComp : GroupMeshes)
+            {
+                float CurrentOpacity = OpacityValues.Contains(MeshComp) ? OpacityValues[MeshComp] : MaxOpacity;
+                float NewOpacity = FMath::FInterpTo(CurrentOpacity, MinOpacity, DeltaSeconds, FadeInSpeed);
+                OpacityValues.Add(MeshComp, NewOpacity);
+                MeshComp->SetScalarParameterValueOnMaterials(FName(TEXT("Opacity")), NewOpacity);
+                MeshesJustOccluded.Add(MeshComp);
+            }
+        }
+        for (auto It = OpacityValues.CreateIterator(); It; ++It)
+        {
+            if (!MeshesJustOccluded.Contains(It.Key()))
+            {
+                float CurrentOpacity = It.Value();
+                float NewOpacity = FMath::FInterpTo(CurrentOpacity, MaxOpacity, DeltaSeconds, FadeOutSpeed);
+            	
+                if (FMath::IsNearlyEqual(NewOpacity, MaxOpacity, 0.01f))
+                {
+                    It.RemoveCurrent();
+                }
+                else
+                {
+                    It.Value() = NewOpacity;
+                    It.Key()->SetScalarParameterValueOnMaterials(FName(TEXT("Opacity")), NewOpacity);
+                }
+            }
+        }
+    }
+    else
+    {
+        for (auto It = OpacityValues.CreateIterator(); It; ++It)
+        {
+            float CurrentOpacity = It.Value();
+            float NewOpacity = FMath::FInterpTo(CurrentOpacity, MaxOpacity, DeltaSeconds, FadeOutSpeed);
+            
+            if (FMath::IsNearlyEqual(NewOpacity, MaxOpacity, 0.01f))
+            {
+                It.RemoveCurrent();
+            }
+            else
+            {
+                It.Value() = NewOpacity;
+                It.Key()->SetScalarParameterValueOnMaterials(FName(TEXT("Opacity")), NewOpacity);
+            }
+        }
+    }
+
```

Comment:
Can we extract this? We have no idea what is does when reading it.  
Why is this in OnPossess? How can it be executed multiple times?   
Missing comments to expain what it does.